pageserver/src/deletion_queue.rs

```diff
@@ -0,0 +1,1299 @@
+            sequence: self.sequence,
+            tenants,
+            size: self.size,
+            validated: self.validated,
+        };
+        self.size = 0;
+        other
+    }
+
+    fn is_empty(&self) -> bool {
+        self.tenants.is_empty()
+    }
+
+    fn len(&self) -> usize {
+        self.size
+    }
+
+    /// Returns true if the push was accepted, false if the caller must start a new
+    /// deletion list.
+    fn push(
+        &mut self,
+        tenant: &TenantId,
+        timeline: &TimelineId,
+        generation: Generation,
+        objects: &mut Vec<RemotePath>,
+    ) -> bool {
+        if objects.is_empty() {
+            // Avoid inserting an empty TimelineDeletionList: this preserves the property
+            // that if we have no keys, then self.objects is empty (used in Self::is_empty)
+            return true;
+        }
+
+        let tenant_entry = self
+            .tenants
+            .entry(*tenant)
+            .or_insert_with(|| TenantDeletionList {
+                timelines: HashMap::new(),
+                generation,
+            });
+
+        if tenant_entry.generation != generation {
+            // Only one generation per tenant per list: signal to
+            // caller to start a new list.
+            return false;
+        }
+
+        let timeline_entry = tenant_entry
+            .timelines
+            .entry(*timeline)
+            .or_insert_with(Vec::new);
+
+        let timeline_remote_path = remote_timeline_path(tenant, timeline);
+
+        self.size += objects.len();
+        timeline_entry.extend(objects.drain(..).map(|p| {
+            p.strip_prefix(&timeline_remote_path)
+                .expect("Timeline paths always start with the timeline prefix")
+                .to_string_lossy()
+                .to_string()
+        }));
+        true
+    }
+
+    fn drain_paths(&mut self) -> Vec<RemotePath> {
+        let mut tenants = HashMap::new();
+        std::mem::swap(&mut tenants, &mut self.tenants);
+
+        let mut result = Vec::new();
+        for (tenant, tenant_deletions) in tenants.into_iter() {
+            for (timeline, timeline_layers) in tenant_deletions.timelines.into_iter() {
+                let timeline_remote_path = remote_timeline_path(&tenant, &timeline);
+                result.extend(
+                    timeline_layers
+                        .into_iter()
+                        .map(|l| timeline_remote_path.join(&PathBuf::from(l))),
+                );
+            }
+        }
+
+        result
+    }
+
+    async fn save(&self, conf: &'static PageServerConf) -> anyhow::Result<()> {
+        let path = conf.deletion_list_path(self.sequence);
+
+        let bytes = serde_json::to_vec(self).expect("Failed to serialize deletion list");
+        tokio::fs::write(&path, &bytes).await?;
+        tokio::fs::File::open(&path).await?.sync_all().await?;
+        Ok(())
+    }
+}
+
+impl std::fmt::Display for DeletionList {
```

Comment:
Shouldn't this be `Debug` instead, contents looks like that.

Reply:
I could go either way.  In pageserver all our strings are either for logging or serialization, so the usual separation into "user facing" (Display) and "programmer facing" (Debug) is fuzzy.  I called this one Display because it's not a literal dump of the entire structure, but a summary.

Reply:
Hmm oki resolving.